src/main/java/de/zalando/aruha/nakadi/service/subscription/zk/CuratorZkSubscriptionClient.java

```diff
@@ -0,0 +1,356 @@
+package de.zalando.aruha.nakadi.service.subscription.zk;
+
+import de.zalando.aruha.nakadi.service.subscription.model.Partition;
+import de.zalando.aruha.nakadi.service.subscription.model.Session;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.locks.InterProcessSemaphoreMutex;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CuratorZkSubscriptionClient implements ZkSubscriptionClient {
+    private static final String STATE_INITIALIZED = "INITIALIZED";
+    private static final Charset CHARSET = Charset.forName("UTF-8");
+
+    private final CuratorFramework curatorFramework;
+    private final InterProcessSemaphoreMutex lock;
+    private final String subscriptionId;
+    private final Logger log;
+
+    private Consumer<Exception> exceptionListener;
+
+    public CuratorZkSubscriptionClient(final String subscriptionId, final CuratorFramework curatorFramework) {
+        this.subscriptionId = subscriptionId;
+        this.curatorFramework = curatorFramework;
+        this.lock = new InterProcessSemaphoreMutex(curatorFramework, "/nakadi/locks/subscription_" + subscriptionId);
+        this.log = LoggerFactory.getLogger("zk." + subscriptionId);
+    }
+
+    public void setExceptionListener(final Consumer<Exception> exceptionListener) {
+        this.exceptionListener = exceptionListener;
+    }
+
+    @Override
+    public void lock(final Runnable function) {
```

Comment:
in fact it does:
1) lock
2) function execution
3) release
I think the name doesn't completely reflects what happens.
Maybe "runLocked" ?